examples/empirical_sinkhorn_div/script.jl

```diff
@@ -0,0 +1,94 @@
+# 
+# While entropy-regularised optimal transport $\operatorname{OT}_{\varepsilon}(\cdot, \cdot)$ is commonly used as a loss function, it suffers from a problem of *bias*: namely that $\nu \mapsto \operatorname{OT}_{\varepsilon}(\mu, \nu)$ is *not* minimised at $\nu = \mu$. 
+#
+# Feydy et al. fix this problem by introducing the *Sinkhorn divergence* between two measures $\mu$ and $\nu$, defined as 
+# ```math
+# \operatorname{S}_{\varepsilon}(\mu, \nu) = \operatorname{OT}_{\varepsilon}(\mu, \nu) - \frac{1}{2} \operatorname{OT}_{\varepsilon}(\mu, \mu) - \frac{1}{2} \operatorname{OT}_{\varepsilon}(\nu, \nu).
+# ```
+# Like the Sinkhorn loss, the Sinkhorn divergence is smooth and convex in both of its arguments. However, the Sinkhorn divergence is unbiased -- i.e. $S_{\varepsilon}(\mu, \nu) = 0$ iff $\mu = \nu$. 
+#
+# Unlike previous examples, here we demonstrate a learning problem similar to Figure 1 of Feydy et al. over *empirical measures*, i.e. measures that have the form $\mu = \frac{1}{N} \sum_{i = 1}^{N} \delta_{x_i}$ where $\delta_x$ is the Dirac delta function at $x$. 
+# 
+# We first load packages. 
+#
+using OptimalTransport
+using ReverseDiff
+using Distributions
+using LinearAlgebra
+using Distances
+using Plots
+using Logging
+using Optim
+
+# As a ground truth distribution, we set $\rho$ to be a Gaussian mixture model with `k = 5` components, equally spaced around a circle, and sample an empirical distribution of size $N$, $\mu \sim \rho$.
+
+k = 3
+d = 2
+θ = π * range(0, 2(1 - 1 / k); length=k)
+μ = 2 * hcat(sin.(θ), cos.(θ))
+ρ = MixtureModel(MvNormal[MvNormal(x, 0.25 * I) for x in eachrow(μ)])
+N = 100
+μ_spt = rand(ρ, N)'
+scatter(μ_spt[:, 1], μ_spt[:, 2]; markeralpha=0.25, title=raw"$\mu$")
+
+# Now, suppose we want to approximate $\mu$ with another empirical distribution $\nu$, i.e. we want to minimise $\nu \mapsto \operatorname{S}_{\varepsilon}(\mu, \nu)$ over possible empirical distributions $\nu$. In this case we have $M$ particles in $\nu$, which we initialise following a Gaussian distribution. 
+M = 100
+ν_spt = rand(M, d);
+# Assign uniform weights to the Diracs in each empirical distribution. 
+μ = fill(1 / N, N)
+ν = fill(1 / M, M);
+
+# Since $\mu$ is fixed, we pre-compute the cost matrix $C_{\mu}$. 
+C_μ = pairwise(SqEuclidean(), μ_spt');
+
+# Define the loss function to minimise, where `x` specifies the locations of the Diracs in $\nu$. 
+function loss(x, ε)
+    C_μν = pairwise(SqEuclidean(), μ_spt', x')
+    C_ν = pairwise(SqEuclidean(), x')
+    return sinkhorn_divergence(μ, ν, C_μν, C_μ, C_ν, ε; maxiter=50)
+end
+# Set entropy regularisation parameter
+ε = 1.0;
+
+# Use ReverseDiff with a precompiled tape and Optim.jl to minimise $\nu \mapsto \operatorname{S}_{\varepsilon}(\mu, \nu)$. Note that this is problem is *not* convex, so we find a local minimium. 
```

Comment:
I don't think this is valid - ReverseDiff with a compiled tape will only yield correct gradients if the function does not contain any branches or if statements. However, the number of iterations in the Sinkhorn algorithm is determined at run time and not fixed.

Reply:
Yes you're absolutely right. Makes sense to me but probably by that time I was copypasting from their tutorial.
A workaround for now would be to set `atol=rtol=0` and fix `maxiter = 50` or something. 
In the long run it might be useful to have an option to set if we want to avoid convergence checks and just run for fixed nb of iterations.